### ISSUE TITLE
Add multi-arg support to PREBUILD/POSTBUILD in DSC file [Defines] section (Bugzilla Bug 670)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#7091]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 670
Date: 2017-08-11T14:17:40+00:00
From: @mdkinney
To:   @mdkinney
CC:   ye.y.wu

Depends on:   #669
Last updated: 2017-11-14T20:55:22+00:00

**Comment by tianocore-issues** [comment]
## Comment 2597
Date: 2017-08-11 14:17:40 +0000
From: @mdkinney

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): @mdkinney

edk2-devel RFC discussion:

https://lists.01.org/pipermail/edk2-devel/2017-August/012883.html

A platform DSC file can provide PREBUILD and POSTBUILD actions in the 
[Defines] section of the DSC file.  The EDK II DSC specification only allows
a single argument for these two defines, which means the PREBUILD and 
POSTBUILD actions are limited to simple batch files, shell scripts, or
applications with no additional arguments.  This also means that a platform
DSC file that uses the PREBUILD/POSTBUILD feature is typically limited to
single OS development environment.

* Expand the PREBUILD and POSTBUILD define syntax to support multiple arguments.
* All arguments are scanned.
* Arguments that are WORKSPACE or PACKAGES_PATH relative file paths are converted
  to absolute paths.  

Supporting multiple arguments allows PREBUILD and POSTBUILD
actions to be implemented in OS independent scripts such as Python. The 
following is an example of a PREBUILD action that runs the python interpreter
on a python script provided in a platform package:

[Defines]
  PREBUILD = python MyPlatformPackage/MyPreBuildScript.py

Supporting multiple arguments also allows additional platform arguments to be
passed into the script:

[Defines]
  PREBUILD = python MyPlatformPackage/MyPreBuildScript.py --FlashSize 4MB

In addition, the EDK II Build Specification states that TARGET, ARCH, and
TOOL_CHAIN_TAG are passed to the PREBUILD and POSTBUILD commands.  The path
to the platform DSC file being build, the path to the Conf directory, and the 
build target form the following set are not provided:

  [all|fds|genc|genmake|clean|cleanall|cleanlib|modules|libraries|run]
  
It is important for the PREBUILD and POSTBUILD actions to receive the same 
context that is provided to the build command. 

Expands the context passed into the PREBUILD and POSTBUILD actions to
include the -p &lt;DSC File&gt;, --conf=&lt;Conf Directory&gt;, and build target
arguments.

There have also been requests to support makefiles in the PREBUILD
and POSTBUILD actions.  Makefiles receive context from system environment
and define values passed in on the command line.  In order to support
makefiles, add a generic python script in /BaseTools/Scripts/RunMakefile.py
This python script converts command line arguments to makefile defines
and runs a makefile specified with the --makefile flag.

Also provide a sample makefile that echos the relevant system environment
variables and defines in BaseTools/Scripts/RunMakefileSample.mak.

An example PREBUIILD action that runs a makefile is as follows:

[Defines]
  PREBUILD = python BaseTools/Script/RunMakefile.py --makefile BaseTools/Scripts/RunMakefileSample.mak

The example above contains 4 arguments.  2 of these arguments are WORKSPACE or
PACKAGES_PATH relative paths that must be converted to absolute paths.  This
is a sample use case for PREBUILD and POSTBUILD define statements that support
multiple arguments and can be used to test this feature and provide a template
makefile for implementing makefiles for PREBUILD and POSTBUILD actions.

**Comment by tianocore-issues** [comment]
## Comment 2610
Date: 2017-08-16 21:04:48 +0000
From: Chris &lt;<ye.y.wu>&gt;

assign to Mike

**Comment by tianocore-issues** [comment]
## Comment 2623
Date: 2017-08-22 21:58:45 +0000
From: @mdkinney

Commits:

https://github.com/tianocore/edk2/commit/af9c4e5e67ab12a77434564ff61c124ef0e5b5dd
https://github.com/tianocore/edk2/commit/c8d385265fa38e0933768fccc8dc0bf5b8725037
https://github.com/tianocore/edk2/commit/ef190542b4dbc4577ef7d1f99209199f0c1b6e66

**Comment by tianocore-issues** [comment]
## Comment 3065
Date: 2017-11-14 20:55:22 +0000
From: @mdkinney

*** Bug 627 has been marked as a duplicate of this bug. ***
